### ISSUE TITLE
Missing battery DesignCapacity and FullChargeCapacity in Windows

### ISSUE THREAD

**Issue by plakak** [sebhildebrandt/systeminformation#539]
Sorry @sebhildebrandt for another battery related issue. It's not like I'm looking for them 🙂  

### Issue:
On my `Dell XPS 15` I don't get `DesignCapacity` and `FullChargeCapacity`. One of my clients also doesn't have those on his newer `XPS 13` as well.

My `wmic`: 
```
BatteryStatus=1
DesignCapacity=
DesignVoltage=11667
EstimatedChargeRemaining=80
FullChargeCapacity=
```

### Possible solution:
After some research the solution that worked for me was to use PowerShell-only class to get those data:

```
PS C:\Users\Alex\systeminformation> (Get-WmiObject -Class BatteryStaticData -Namespace ROOT/WMI).DesignedCapacity            
83995
PS C:\Users\Alex\systeminformation> (Get-WmiObject -Class BatteryFullChargedCapacity -Namespace ROOT/WMI).FullChargedCapacity
44232
```

I'm not sure how much backward compatible is this but I used those as a fallback if no data is available from `wmic` and it works correctly:
```
exec('(Get-WmiObject -Class BatteryStaticData -Namespace ROOT/\WMI).DesignedCapacity', {'shell':'powershell.exe'})
exec('(Get-WmiObject -Class BatteryFullChargedCapacity -Namespace ROOT/\WMI).FullChargedCapacity' {'shell':'powershell.exe'}),
```

Environment (please complete the following information):
systeminformation package version: 5.6.21
OS: Windows 10
Hardware: Dell XPS 15 9550

**Comment by sebhildebrandt** [comment]
@plakak Thank you for reporting this! Don’t worry. I am also interested in making this as stable as possible. Will have a look at it and provide a solution tomorrow.

**Comment by sebhildebrandt** [comment]
@plakak can you try one more thing: When we relate this to #537 then it would be good to get the output of 

```
(Get-WmiObject -Class BatteryStaticData -Namespace ROOT/WMI).DesignedCapacity 
(Get-WmiObject -Class BatteryFullChargedCapacity -Namespace ROOT/WMI).FullChargedCapacity
```

on the machines you where talking about in the other issue.
I would like to handle both issues correctly in one patch.

Thank you in advance.

**Comment by plakak** [comment]
@sebhildebrandt  Here it is 🙂 

![image (3)](https://user-images.githubusercontent.com/13939486/119141543-ac3b2880-ba45-11eb-8b92-feefa3c08069.png)

**Comment by sebhildebrandt** [comment]
@plakak perfect. Thank you so much!!

**Comment by plakak** [comment]
@sebhildebrandt any ETA when to expect this? I'm wondering should I make another release with my own fixes or wait for the official ones if they are close enough 🙂 

Cheers!

**Comment by sebhildebrandt** [comment]
Sorry @plakak did not had time ... hope to finish this by tomorrow.

**Comment by sebhildebrandt** [comment]
@plakak I now pushed a code to GitHub that should integrate this additional possibility to get designed and full charged capacity. This code change also fixes #537 . So it would be nice if you can grab the code from GitHub and test it on the specific machines where you had problems. Please report back, if you see any problems.

**Comment by plakak** [comment]
@sebhildebrandt I tried on a Dell laptop and unfortunately it didn't work. What's more, I'm not getting any battery data now. I did quick debugging and:

1. Even thought I'm on one battery device part `for (let i = 0; i < parts.length; i++) {` gets fired 3 times. I have values  (on first pass) in `designCapacity` and `fullChargeCapacity` variables but the `parsed` has status `-1`. On the second and third pass, both variables have 0 while parsed is still -1.
2. I went up and in `parseWinBatteryPart` I get `2` in `status` variable - so correct, but `lines.join('').toLowerCase()` returns ` timetofullcharge=p-64jpsdrn32_computersystem` which leads to issue with the if condition.

**Comment by sebhildebrandt** [comment]
@plakak thank you, need to make investigation. Unfortunately I do not have a laptop with windows installed. So therefore this was kind of a blind shot ... will come back with another try ...

**Comment by plakak** [comment]
@sebhildebrandt Maybe I can build based on your code now, implement parts of my solution using your approach? I could probably sit to it in the evening and see if I come up with something you would approve? I struggled with getting the right parsing myself, and I do have a windows machine :)

**Comment by sebhildebrandt** [comment]
@plakak that would be great ... we need to be sure to keep node 4 compatibility (no map, spread, ...) ;-)

**Comment by plakak** [comment]
@sebhildebrandt just to confirm - https://nodejs.org/en/blog/release/v4.0.0/ based on that `map` and `reduce` should be fine. Should I avoid that regardless or those two are fine? 🙂

**Comment by plakak** [comment]
Ok, got a version that works on Dell without any "fancy" features 😄 I'll pass it to the "ThinkPad guy" to test if that also works correctly there 👍🏻

**Comment by sebhildebrandt** [comment]
@plakak WOW great! Any possibility to also review it on my side?

**Comment by plakak** [comment]
@sebhildebrandt of course, just want to make sure it's working on all machines. I'll create a PR for you as soon as I have this confirmation :)

**Comment by sebhildebrandt** [comment]
@plakak perfect, totally agree!

**Comment by plakak** [comment]
@sebhildebrandt PR is ready. Any thoughts if it's OK and if yes when you'd be merging it to master? I need to plan some dates for the next release in my project and I'm curious 🙂

**Comment by sebhildebrandt** [comment]
@plakak I had a look on it ... so you changed the way how to split the `parts` ... I just wondered why, because this is how it works also for all other (wmic based windows) functions. And to be consistent to the rest of the code, maybe we can stick with this? I will try to make a modification ...

**Comment by plakak** [comment]
It's for separate code for multiple batteries. I don't know if anywhere else you have result presented that way as with multiple batteries. It just takes each part and pushes it to one array, if there is empty line or multiple empty lines it creates new array and then pushes parts after that empty line into second array etc. If one would have 5 batteries there would be 5 arrays with parts. So this code just separate one long 'parts' that may contain data for multiple batteries into separate 'parts' for each battery. And then we don't have to do `let lines = parts[i].split('\r\n');` because we already have exactly that array as a battery `let lines = batteries[i]`. So basically it does the same that your code but just aggregates parts per battery a bit earlier. I'm not a fan of `lines.join('').toLowerCase().indexOf('BatteryStatus') >= 0` that much -  I found that the solution of having batteries split beforehand and your helper function getting always correct lines to parse a bit cleaner to my liking.

TLDR; Basically it uses all your code apart the fact that it does not try to find where one battery ends and second starts via regexp or looking on string, rather it splits in array of batteries a bit sooner and then uses lines from each battery.

**Comment by sebhildebrandt** [comment]
@plakak what I mean: when I do a `wmic ... get /value` and get multiple e.g. batteries, then first I do split the parts with `split(/\n\s*\n/);`. Then I walk through each if the result parts (which would then be each battery) ...:

```
let parts = data.results[0].split(/\n\s*\n/);
if (parts && parts.length) {
for (let i = 0; i < parts.length; i++) {
	let lines = parts[i].split('\r\n');

	... proceed with each part ...
	
}
``` 

This normally works fine (so in your example case there should also be 5 parts), but you have changed it and I am not sure why ... anyway, then I will just take your code ...

**Comment by sebhildebrandt** [comment]
@plakak will do it today ;-) to give a time frame ...

**Comment by plakak** [comment]
@sebhildebrandt I actually started from debugging it from the code you wrote but I couldn't make it work. I don't remember exactly why but I think it was because the response is looking a bit different in terms of linebrakes
```
\r\r\n\r\r\nBatteryStatus=\r\r\nDesignCapacity=\r\r\nDesignVoltage=11599\r\r\nnFullChargeCapacity=\r\r\n\r\r\n\r\r\n\r\r\n
```
then second battery etc. So I just wrote code that ignores all the empty lines and cares only for the "right" lines (`/\S/.test(value)`) - no matter if this is `\r\r\n`, `\r\n` etc as it was difficult to find the correct linebrake for all situations + I think it might be a bit different on each system (win 10, 8 etc).

**Comment by sebhildebrandt** [comment]
@plakak ... first of all: I really have to thank you for putting so much effort to improving this!!! And secondly: I do agree, This works fine this way... I just wrote the code as like all the other parts I did in other sections ;-) But no worries. I will just take it how it is now and release a new version today. Best regards!!!

**Comment by sebhildebrandt** [comment]
@plakak can you try the new version 5.7.6 ... should now work as expected.

**Comment by plakak** [comment]
@sebhildebrandt I'll check that in the evening - thanks!

**Comment by sebhildebrandt** [comment]
@plakak take your time ... thank you!

**Comment by plakak** [comment]
@sebhildebrandt  Dell laptop works correctly. In terms of Thinkpad it will take a while until I can get the feedback but I would assume, based on previous tests it should run smoothly as well.

You dIdn't use my PR after all? It was open to add any additional changes to it. To tell the truth I'm a bit disappointed not to be included as a contributor after finding and providing fixes/way of fixing for those two issues. Getting thought all the documentation to find the powershell workaround was not as easy as it sounds as well and costed me quite some time...

**Comment by sebhildebrandt** [comment]
@plakak sorry! My fault ... while checking  my repo  I just copied it ... now merged the pull request, you are now official contributor and I anyway will add you as a contributor to the official docs ...
